# Load libraries

In [11]:
import pandas as pd 
import numpy as np
import re

# Load datasets

In [12]:
df1 = pd.read_csv("data/ALLARMI.csv")
df2 = pd.read_csv("data/TIPOLOGIA_VIAGGIATORE.csv")

In [13]:
df1

,OCCORRENZE,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,CITTA_PARTENZA,...,ZONA,TOT,MOTIVO_ALLARME,note_operatore,flag_rischio,Paese Partenza,CODICE PAESE ARR,3zona,paese%arr,tot voli
0,Voli con Allarmi,FCO,IST,2024,01,2024-01-30 09:15:00,Fiumicino,Havalimani,Roma,Istanbul,...,5,1,Manuale,NaN,NaN,Turchia,ITA,5,Italia,1
1,Viaggiatori con Allarmi,CIA,STN,2024,02,2024-02-03 13:15:00,Ciampino,Stansted,Roma,Londra,...,5,5,Manuale,NaN,NaN,Regno Unito,ITA,5,Italia,5
2,Viaggiatori entrati nel Sistema,FCO,LHR,2024,01,2024-01-15 08:45:00,Fiumicino,London Heathrow,Roma,Londra,...,5,110,TSC,NaN,NaN,Regno Unito,ITA,5,Italia,110
3,Voli con Allarmi,MXP,LHR,2024,02,2024-02-02 08:40:00,Malpensa,London Heathrow,Milano,Londra,...,2,1,SDI,NaN,NaN,Regno Unito,ITA,2,Italia,1
4,Viaggiatori con Allarmi,PSA,BRS,2024,02,2024-02-16 12:50:00,Galileo Galilei,Bristol,Pisa,Bristol,...,8,2,INTERPOL,NaN,NaN,Regno Unito,ITA,8,Italia,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5075,Nulla a procedere SDI,MXP,LTN,2023,01,2024-01-19 17:30:00,Malpensa,London Luton,Milano,Londra,...,2,5,TSC,NaN,NaN,Regno Unito,ITA,2,Italia,5
5076,Voli disponibili in ingresso al Sistema,BLQ,RAK,2024,02,FEB 06 2024,Guglielmo Marconi,Menara,Bologna,Marrakech,...,8,1,TSC,NaN,NaN,Marocco,ITA,8,Italia,1
5077,Voli con Allarmi,FCO,DOH,2024,01,2024-01-12 15:35:00,Fiumicino,Hamad International,Roma,Doha,...,5,1,NSIS,NaN,NaN,Qatar,ITA,5,Italia,1
5078,"Voli solo visualizzati, ma NON investigati",LIN,LHR,2024,01,2024-01-13 11:50:00,Linate,London Heathrow,Milano,Londra,...,2,1,TSC,NaN,NaN,Regno Unito,ITA,2,Italia,1


In [14]:
df2

,NAZIONALITA,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,GIORNO_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,...,COMPAGNIA_AEREA,NUMERO_VOLO,ESITO_CONTROLLO,note_operatore,codice_rischio,Tipo Documento,FASCIA ETA,3nazionalita,compagnia%aerea,num volo
0,ALB,NAP,DUR,2024,02,13,2024-02-13 07:30:00,Napoli Capodichino,King Shaka International,Napoli,...,Fly Dubai,FZ1681,RESPINTO,NaN,NaN,Passaporto,N.D.,ALB,Fly Dubai,FZ1681
1,NaN,FCO,JFK,2024,01,22,2024-01-22 16:35:00,Fiumicino,John F Kennedy International,Roma,...,ITA Airways,AZ0609,NaN,NaN,NaN,Carta d'identità,18-30,ALB,ITA Airways,AZ0609
2,ALB,TSF,TIA,2024,02,4,2024-02-04 20:10:00,Treviso-Sant'Angelo,Rinas Mother Teresa,Treviso,...,Ryanair DAC,FR8400,SEGNALATO,NaN,NaN,N.D.,31-45,ALB,Ryanair DAC,FR8400
3,AFG,FCO,IST,2024,01,25,2024-01-25 13:05:00,Fiumicino,Havalimani,Roma,...,Turkish Airlines,TK1865,NaN,NaN,NaN,N.D.,61+,AFG,Turkish Airlines,TK1865
4,ALB,BGY,MLE,2024,02,13,FEB 13 2024,Orio al Serio,Male International,Bergamo,...,Fly Dubai,FZ1571,SEGNALATO,NaN,NaN,Permesso di soggiorno,46-60,ALB,Fly Dubai,FZ1571
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5090,ALB,BLQ,TIA,2024,02,15,2024-02-18 09:10:00,Guglielmo Marconi,Rinas Mother Teresa,Bologna,...,Ryanair DAC,FR8399,NaN,NaN,NaN,N.D.,0-17,ALB,Ryanair DAC,FR8399
5091,ALB,FCO,TIA,2024,01,10,2024-01-10 06:10:00,Fiumicino,Rinas Mother Teresa,Roma,...,Wizz Air Malta,W43857,NaN,NaN,NaN,Passaporto,46-60,ALB,Wizz Air Malta,W43857
5092,ALB,FCO,JED,2024,01,31,31/01/2024 22:15:00,Fiumicino,King Abdulaziz International,Roma,...,Wizz Air Malta,w48350,FERMATO,NaN,NaN,N.D.,31-45,ALB,Wizz Air Malta,w48350
5093,ALB,BGY,TIA,2024,01,1,2024-01-01 08:25:00,Orio al Serio,Rinas Mother Teresa,Bergamo,...,Ryanair DAC,FR8350,IN ATTESA,NaN,NaN,Carta d'identità,18-30,ALB,Ryanair DAC,FR8350


# First look at the datasets

In [15]:
df1.shape

(5080, 24)

In [16]:
df1.head(5)

,OCCORRENZE,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,CITTA_PARTENZA,...,ZONA,TOT,MOTIVO_ALLARME,note_operatore,flag_rischio,Paese Partenza,CODICE PAESE ARR,3zona,paese%arr,tot voli
0,Voli con Allarmi,FCO,IST,2024,01,2024-01-30 09:15:00,Fiumicino,Havalimani,Roma,Istanbul,...,5,1,Manuale,NaN,NaN,Turchia,ITA,5,Italia,1
1,Viaggiatori con Allarmi,CIA,STN,2024,02,2024-02-03 13:15:00,Ciampino,Stansted,Roma,Londra,...,5,5,Manuale,NaN,NaN,Regno Unito,ITA,5,Italia,5
2,Viaggiatori entrati nel Sistema,FCO,LHR,2024,01,2024-01-15 08:45:00,Fiumicino,London Heathrow,Roma,Londra,...,5,110,TSC,NaN,NaN,Regno Unito,ITA,5,Italia,110
3,Voli con Allarmi,MXP,LHR,2024,02,2024-02-02 08:40:00,Malpensa,London Heathrow,Milano,Londra,...,2,1,SDI,NaN,NaN,Regno Unito,ITA,2,Italia,1
4,Viaggiatori con Allarmi,PSA,BRS,2024,02,2024-02-16 12:50:00,Galileo Galilei,Bristol,Pisa,Bristol,...,8,2,INTERPOL,NaN,NaN,Regno Unito,ITA,8,Italia,2


In [17]:
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 5080 entries, 0 to 5079
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   OCCORRENZE             5080 non-null   str  
 1   AREOPORTO_ARRIVO       5080 non-null   str  
 2   AREOPORTO_PARTENZA     5080 non-null   str  
 3   ANNO_PARTENZA          5080 non-null   str  
 4   MESE_PARTENZA          5080 non-null   str  
 5   DATA_PARTENZA          5080 non-null   str  
 6   DESCR_AEREOPORTO_ARR   5080 non-null   str  
 7   DESCR_AEREOPORTO_PART  4971 non-null   str  
 8   CITTA_ARR              5080 non-null   str  
 9   CITTA_PARTENZA         4979 non-null   str  
 10  CODICE_PAESE_ARR       5080 non-null   str  
 11  CODICE_PAESE_PART      5026 non-null   str  
 12  PAESE_ARR              5080 non-null   str  
 13  PAESE_PART             5006 non-null   str  
 14  ZONA                   5080 non-null   str  
 15  TOT                    5080 non-null   str  
 16 

In [18]:
# ── df1 – Normalizzazione testo (lowercase + strip) ──────────────────────────
df1 = df1.map(lambda x: x.strip().lower() if isinstance(x, str) else x)

In [19]:
# ── df1 – OCCORRENZE ─────────────────────────────────────────────────────────
print(f"Valori unici: {df1['OCCORRENZE'].nunique()}")
print(f"Valori nulli: {df1['OCCORRENZE'].isnull().sum()}")
print()
print(df1['OCCORRENZE'].value_counts(dropna=False).to_string())

Valori unici: 30
Valori nulli: 0

OCCORRENZE
viaggiatori entrati nel sistema                 626
voli disponibili in ingresso al sistema         588
voli solo visualizzati, ma non investigati      572
voli investigati (sdi/nsis - interpol - tsc)    543
viaggiatori investigati                         528
allarmi generati da sdi/nsis                    513
voli con allarmi                                507
viaggiatori con allarmi                         468
nulla a procedere sdi                           251
allarmi chiusi                                  245
allarmi rilevanti                               112
errata segnalazione sdi                          25
allarmi generati da bcs                          21
allarmi chiusi con azione (cc.xx)                14
notifica atti/provv                              10
allarmi generati da interpol                      8
allarmi non chiusi                                7
mancato aggiornamento sdi                         6
nulla a procedere n

In [20]:
# ── df1 – Raggruppamento OCCORRENZE ──────────────────────────────────────────

occorrenze_map = {
    'viaggiatori entrati nel sistema'              : 'viaggiatori entrati',
    'voli disponibili in ingresso al sistema'      : 'voli disponibili',
    'voli solo visualizzati, ma non investigati'   : 'voli solo visualizzati',
    'voli investigati (sdi/nsis - interpol - tsc)' : 'voli investigati',
    'viaggiatori investigati'                      : 'viaggiatori investigati',
    'voli con allarmi'                             : 'voli con allarmi',
    'viaggiatori con allarmi'                      : 'viaggiatori con allarmi',
    'allarmi generati da sdi/nsis'                 : 'allarmi generati',
    'allarmi generati da bcs'                      : 'allarmi generati',
    'allarmi generati da interpol'                 : 'allarmi generati',
    'allarmi generati'                             : 'allarmi generati',
    'allarmi chiusi'                               : 'allarmi chiusi',
    'allarmi chiusi con azione (cc.xx)'            : 'allarmi chiusi',
    'allarmi rilevanti'                            : 'allarmi rilevanti',
    'allarmi non chiusi'                           : 'allarmi non chiusi',
    'nulla a procedere sdi'                        : 'esito negativo',
    'nulla a procedere nsis'                       : 'esito negativo',
    'nulla a procedere int'                        : 'esito negativo',
    'errata segnalazione sdi'                      : 'esito negativo',
    'errata segnalazione nsis'                     : 'esito negativo',
    'errata segnalazione bcs'                      : 'esito negativo',
    'respinto/a'                                   : 'esito negativo',
    'mancato aggiornamento sdi'                    : 'mancato aggiornamento',
    'mancato aggiornamento schengen nsis'          : 'mancato aggiornamento',
    'notifica atti/provv'                          : 'altro',
    'inammissibilita schengen - art.24'            : 'altro',
    'altro'                                        : 'altro',
    '???'                                          : np.nan,
    'n/c'                                          : np.nan,
    'ali armatt'                                   : np.nan,
}

df1['OCCORRENZE'] = df1['OCCORRENZE'].map(occorrenze_map)

print(f"\nValori nulli: {df1['OCCORRENZE'].isnull().sum()}")
print(f"\nDistribuzione:")
print(df1['OCCORRENZE'].value_counts(dropna=False).to_string())


Valori nulli: 3

Distribuzione:
OCCORRENZE
viaggiatori entrati        626
voli disponibili           588
voli solo visualizzati     572
allarmi generati           547
voli investigati           543
viaggiatori investigati    528
voli con allarmi           507
viaggiatori con allarmi    468
esito negativo             296
allarmi chiusi             259
allarmi rilevanti          112
altro                       16
mancato aggiornamento        8
allarmi non chiusi           7
NaN                          3


In [21]:
# ── df1 – AREOPORTO_ARRIVO ───────────────────────────────────────────────────
print(f"Valori unici: {df1['AREOPORTO_ARRIVO'].nunique()}")
print(f"Valori nulli: {df1['AREOPORTO_ARRIVO'].isnull().sum()}")
print()
print(df1['AREOPORTO_ARRIVO'].value_counts(dropna=False).to_string())

Valori unici: 31
Valori nulli: 0

AREOPORTO_ARRIVO
fco    1283
mxp    1194
bgy     471
vce     359
blq     329
nap     204
lin     175
cia     166
trn     154
psa     140
cta     115
vrn     110
tsf      83
flr      69
bri      57
pmo      26
aoi      24
goa      20
peg      16
psr      16
trs      14
rmi      12
suf       9
pmf       7
cag       7
bds       6
cuf       6
ciy       3
olb       2
bzo       2
vbs       1


In [22]:
# ── df1 – AREOPORTO_PARTENZA ─────────────────────────────────────────────────
print(f"Valori unici: {df1['AREOPORTO_PARTENZA'].nunique()}")
print(f"Valori nulli: {df1['AREOPORTO_PARTENZA'].isnull().sum()}")
print()
print(df1['AREOPORTO_PARTENZA'].value_counts(dropna=False).to_string())

Valori unici: 112
Valori nulli: 0

AREOPORTO_PARTENZA
tia    621
stn    506
lgw    402
lhr    393
ist    302
dxb    153
cai    134
saw    134
cmn    130
man    129
doh    106
rak    106
jfk    104
ssh     96
tun     86
beg     81
auh     78
lcy     74
ewr     63
ltn     60
jed     52
edi     52
gru     51
brs     47
tlv     46
amm     46
add     46
rmo     43
bhx     41
pek     35
evn     32
spx     31
eze     31
pvg     30
icn     29
del     28
tpe     25
alg     25
kwi     23
sin     22
mct     21
ruh     21
ema     18
skp     17
mle     17
szx     17
gig     17
atl     16
mia     16
kiv     15
gyd     15
ncl     15
rmf     15
shj     15
wnz     13
yul     13
bah     13
ika     12
dss     12
kut     12
mex     11
tgd     11
phl     11
fez     10
mba     10
gla     10
hnd     10
lrm      9
hrg      9
sid      9
bqh      9
hkg      8
tng      8
fab      8
dfw      8
prn      8
dwc      8
bey      7
lpl      7
hgh      6
yyz      6
tbs      6
iad      6
adb      6
znz      5
rba      5


In [23]:
# ── df1 – ANNO_PARTENZA ──────────────────────────────────────────────────────
print(f"Valori unici: {df1['ANNO_PARTENZA'].nunique()}")
print(f"Valori nulli: {df1['ANNO_PARTENZA'].isnull().sum()}")
print()
print(df1['ANNO_PARTENZA'].value_counts(dropna=False).to_string())

Valori unici: 5
Valori nulli: 0

ANNO_PARTENZA
2024         4677
2023          257
24             52
anno 2024      52
2024.          42


In [24]:
# ── df1 – ANNO_PARTENZA ──────────────────────────────────────────────────────

def normalizza_anno(val):
    if pd.isna(val): return np.nan
    val = str(val).strip()
    match = re.search(r'\b(20\d{2})\b', val)
    if match: return int(match.group(1))
    match = re.search(r'\b(\d{2})\b', val)
    if match: return int('20' + match.group(1))
    return np.nan

df1['ANNO_PARTENZA'] = df1['ANNO_PARTENZA'].apply(normalizza_anno)

print(df1['ANNO_PARTENZA'].value_counts(dropna=False).to_string())

ANNO_PARTENZA
2024    4823
2023     257


In [25]:
# ── df1 – MESE_PARTENZA ──────────────────────────────────────────────────────
print(f"Valori unici: {df1['MESE_PARTENZA'].nunique()}")
print(f"Valori nulli: {df1['MESE_PARTENZA'].isnull().sum()}")
print()
print(df1['MESE_PARTENZA'].value_counts(dropna=False).to_string())

Valori unici: 14
Valori nulli: 0

MESE_PARTENZA
01     2453
02     2290
gen      81
feb      63
12       26
07       23
06       22
04       21
11       20
09       18
08       16
05       16
03       16
10       15


In [26]:
# fix MESE_PARTENZA
mesi_map = {"GEN": "01", "FEB": "02"}
df1["MESE_PARTENZA"] = df1["MESE_PARTENZA"].replace(mesi_map)

# verifica
print(df1["MESE_PARTENZA"].value_counts(dropna=False).sort_index().to_string())

MESE_PARTENZA
01     2453
02     2290
03       16
04       21
05       16
06       22
07       23
08       16
09       18
10       15
11       20
12       26
feb      63
gen      81


In [27]:
# ── df1 – DATA_PARTENZA ──────────────────────────────────────────────────────
print(f"Valori unici: {df1['DATA_PARTENZA'].nunique()}")
print(f"Valori nulli: {df1['DATA_PARTENZA'].isnull().sum()}")
print()
print(df1['DATA_PARTENZA'].value_counts(dropna=False).to_string())

Valori unici: 3810
Valori nulli: 0

DATA_PARTENZA
2024-02-04 06:10:00    7
2024-02-29 16:40:00    6
2024-01-31 18:00:00    6
2024-01-19 08:55:00    5
2024-01-22 13:00:00    5
2024-02-08 10:00:00    5
2024/02/26             5
2024-01-04 11:30:00    5
2024-01-06 08:00:00    5
2024-01-07 18:00:00    5
2024-02-18 06:50:00    5
2024-02-09 12:50:00    4
05.01.2024             4
2024/02/25             4
gen 06 2024            4
2024-01-05 13:40:00    4
feb 18 2024            4
2024-01-30 12:35:00    4
18.02.2024             4
2024-01-11 14:05:00    4
2024-01-22 07:30:00    4
2024-02-19 09:00:00    4
2024-01-07 17:00:00    4
gen 27 2024            4
2024-02-10 22:20:00    4
gen 14 2024            4
2024-01-12 15:35:00    4
17-01-24               4
2024-02-16 17:30:00    4
2024-02-18 13:05:00    4
2024-01-01 08:00:00    4
2024-02-02 09:15:00    4
2024-01-05 19:05:00    4
2024-01-27 18:05:00    4
23-02-24               4
2024-01-24 08:00:00    4
2024-02-09 18:05:00    4
2024-01-21 17:50:00    4


In [28]:
import pandas as pd
import re

mesi_ita = {
    "GEN": "01", "FEB": "02", "MAR": "03", "APR": "04",
    "MAG": "05", "GIU": "06", "LUG": "07", "AGO": "08",
    "SET": "09", "OTT": "10", "NOV": "11", "DIC": "12"
}

def normalizza_data(val):
    val = str(val).strip()

    # "GEN 06 2024" → "2024-01-06 00:00:00"
    for ita, num in mesi_ita.items():
        if val.startswith(ita + " "):
            parts = val.split(" ")
            return f"{parts[2]}-{num}-{parts[1].zfill(2)} 00:00:00"

    # "2024-01-05T12:35:00" → "2024-01-05 12:35:00"
    val = val.replace("T", " ")

    # "2024/02/26" o "2024/02/26 09:00:00" → "2024-02-26 00:00:00"
    if re.match(r"\d{4}/\d{2}/\d{2}", val):
        val = val.replace("/", "-")
        if len(val) == 10:
            val = val + " 00:00:00"

    # "05.01.2024" → "2024-01-05 00:00:00"
    match = re.match(r"(\d{2})\.(\d{2})\.(\d{4})", val)
    if match:
        return f"{match.group(3)}-{match.group(2)}-{match.group(1)} 00:00:00"

    # "04-01-24" → "2024-01-04 00:00:00"
    match = re.match(r"(\d{2})-(\d{2})-(\d{2})$", val)
    if match:
        return f"20{match.group(3)}-{match.group(2)}-{match.group(1)} 00:00:00"

    # "22/01/2024 15:45:00" → "2024-01-22 15:45:00"
    match = re.match(r"(\d{2})/(\d{2})/(\d{4})\s(\d{2}:\d{2}:\d{2})", val)
    if match:
        return f"{match.group(3)}-{match.group(2)}-{match.group(1)} {match.group(4)}"

    return val

df1["DATA_PARTENZA"] = df1["DATA_PARTENZA"].apply(normalizza_data)
df1["DATA_PARTENZA"] = pd.to_datetime(df1["DATA_PARTENZA"], errors="coerce")

# verifica
print("NaT rimasti:", df1["DATA_PARTENZA"].isnull().sum())
print("Range date:", df1["DATA_PARTENZA"].min(), "→", df1["DATA_PARTENZA"].max())
print("\nEsempi:")
print(df1["DATA_PARTENZA"].head(5).to_string())

NaT rimasti: 130
Range date: 2023-12-31 00:00:00 → 2024-02-29 22:20:00

Esempi:
0   2024-01-30 09:15:00
1   2024-02-03 13:15:00
2   2024-01-15 08:45:00
3   2024-02-02 08:40:00
4   2024-02-16 12:50:00


In [29]:
# ── df1 – DESCR_AEREOPORTO_ARR ──────────────────────────────────────────────────────
print(f"Valori unici: {df1['DESCR_AEREOPORTO_ARR'].nunique()}")
print(f"Valori nulli: {df1['DESCR_AEREOPORTO_ARR'].isnull().sum()}")
print()
print(df1['DESCR_AEREOPORTO_ARR'].value_counts(dropna=False).to_string())

Valori unici: 31
Valori nulli: 0

DESCR_AEREOPORTO_ARR
fiumicino               1283
malpensa                1194
orio al serio            471
venezia marco polo       359
guglielmo marconi        329
napoli capodichino       204
linate                   175
ciampino                 166
sandro pertini           154
galileo galilei          140
fontanarossa             115
verona villafranca       110
treviso-sant'angelo       83
firenze peretola          69
palese                    57
punta raisi               26
falconara                 24
cristoforo colombo        20
sant egidio               16
liberi                    16
ronchi dei legionari      14
miramare                  12
lamezia terme              9
parma                      7
elmas                      7
papola casale              6
levaldigi                  6
comiso                     3
costa smeralda             2
bolzano                    2
brescia montichiari        1


In [30]:
# ── df1 – DESCR_AEREOPORTO_PART ──────────────────────────────────────────────────────
print(f"Valori unici: {df1['DESCR_AEREOPORTO_PART'].nunique()}")
print(f"Valori nulli: {df1['DESCR_AEREOPORTO_PART'].isnull().sum()}")
print()
print(df1['DESCR_AEREOPORTO_PART'].value_counts(dropna=False).to_string())

Valori unici: 119
Valori nulli: 109

DESCR_AEREOPORTO_PART
rinas mother teresa                                                  584
stansted                                                             476
london gatwick                                                       382
london heathrow                                                      370
havalimani                                                           286
dubai international                                                  142
cairo international                                                  127
sabiha gokcen                                                        127
manchester                                                           122
mohamed v                                                            120
NaN                                                                  109
hamad international                                                  103
menara                                                           

## DESCR_AEREOPORTO_PART

La colonna conteneva 3 tipi di problemi:

1. **167 righe con spazi extra** — risolto con `.str.strip()`

2. **83 valori placeholder** — valori come `"n.d."`, `"nd"`, `"unknown"`, `"null"` 
   che non contengono informazione reale ma erano scritti come testo invece di 
   essere lasciati vuoti. Convertiti in `pd.NA` con `.replace()`.

3. **109 null originali** — celle vuote già presenti nel dataset.

I punti 2 e 3 hanno portato il totale dei null a 192 (3.8% delle righe).
Tutti e 192 sono stati **recuperati** utilizzando il codice IATA presente nella 
colonna `AREOPORTO_PARTENZA`: tramite una lookup table IATA → nome aeroporto 
per esteso, ogni null è stato sostituito con il nome corretto in lowercase 
(es. `"STN"` → `"stansted"`, `"LHR"` → `"london heathrow"`).

**Risultato finale: 0 null, 133 valori unici, tutto in lowercase.**

In [32]:
import pandas as pd

# lookup IATA → nome aeroporto per esteso (lowercase)
iata_lookup = {
    "TIA": "rinas mother teresa",
    "STN": "stansted",
    "LGW": "london gatwick",
    "LHR": "london heathrow",
    "IST": "havalimani",
    "DXB": "dubai international",
    "CAI": "cairo international",
    "SAW": "sabiha gokcen",
    "MAN": "manchester",
    "CMN": "mohamed v",
    "DOH": "hamad international",
    "RAK": "menara",
    "JFK": "john f kennedy international",
    "SSH": "sharm el-sheikh international",
    "TUN": "carthage",
    "AUH": "abu dhabi international",
    "BEG": "nikola tesla",
    "LCY": "london city",
    "LTN": "london luton",
    "EWR": "newark liberty international",
    "EZE": "ministro pistarini international",
    "ADD": "bole international",
    "MEX": "benito juarez international",
    "JED": "king abdulaziz international",
    "EDI": "edinburgh",
    "GRU": "guarulhos international",
    "NCL": "newcastle",
    "PEK": "beijing capital international",
    "RMO": "chisinau international",
    "BHX": "birmingham",
    "PVG": "pudong international",
    "HKG": "hong kong international",
    "MCT": "muscat international",
    "MLE": "velana international",
    "RBA": "sale",
    "EVN": "zvartnots international",
    "TBS": "tbilisi international",
    "EMA": "east midlands",
    "RUH": "king khalid international",
    "SHJ": "sharjah international",
    "GLA": "glasgow",
    "ICN": "incheon international",
    "RMF": "marsa alam international",
    "TNG": "ibn batouta",
    "ANU": "v c bird international",
    "SID": "amilcar cabral international",
    "SPX": "sphinx international",
    "IKA": "imam khomeini international",
    "PRN": "pristina international",
    "MBA": "moi international",
    "SZX": "shenzhen bao an international",
    "ATQ": "sri guru ram dass jee international",
    "MIA": "miami international",
    "BQH": "biggin hill",
    "TPE": "taoyuan international",
    "SIN": "changi international",
    "TLV": "ben gurion international",
    "AMM": "queen alia international",
    "BRS": "bristol",
}

# 1. strip + lowercase
df1["DESCR_AEREOPORTO_PART"] = df1["DESCR_AEREOPORTO_PART"].str.lower().str.strip()

# 2. placeholder → NaN
placeholders = ["n.d.", "nd", "unknown", "null"]
df1["DESCR_AEREOPORTO_PART"] = df1["DESCR_AEREOPORTO_PART"].replace(placeholders, pd.NA)

# 3. recupera null dal codice IATA
null_mask = df1["DESCR_AEREOPORTO_PART"].isnull()
df1.loc[null_mask, "DESCR_AEREOPORTO_PART"] = (
    df1.loc[null_mask, "AREOPORTO_PARTENZA"]
    .str.strip().str.upper()
    .map(iata_lookup)
)

# verifica
print(f"Null rimasti: {df1['DESCR_AEREOPORTO_PART'].isnull().sum()}")
print(f"Valori unici: {df1['DESCR_AEREOPORTO_PART'].nunique()}")

Null rimasti: 0
Valori unici: 133


In [158]:
df2.shape

(5095, 33)

In [159]:
df2.head(5)

,NAZIONALITA,AREOPORTO_ARRIVO,AREOPORTO_PARTENZA,ANNO_PARTENZA,MESE_PARTENZA,GIORNO_PARTENZA,DATA_PARTENZA,DESCR_AEREOPORTO_ARR,DESCR_AEREOPORTO_PART,CITTA_ARR,...,COMPAGNIA_AEREA,NUMERO_VOLO,ESITO_CONTROLLO,note_operatore,codice_rischio,Tipo Documento,FASCIA ETA,3nazionalita,compagnia%aerea,num volo
0,ALB,NAP,DUR,2024,02,13,2024-02-13 07:30:00,Napoli Capodichino,King Shaka International,Napoli,...,Fly Dubai,FZ1681,RESPINTO,NaN,NaN,Passaporto,N.D.,ALB,Fly Dubai,FZ1681
1,NaN,FCO,JFK,2024,01,22,2024-01-22 16:35:00,Fiumicino,John F Kennedy International,Roma,...,ITA Airways,AZ0609,NaN,NaN,NaN,Carta d'identità,18-30,ALB,ITA Airways,AZ0609
2,ALB,TSF,TIA,2024,02,4,2024-02-04 20:10:00,Treviso-Sant'Angelo,Rinas Mother Teresa,Treviso,...,Ryanair DAC,FR8400,SEGNALATO,NaN,NaN,N.D.,31-45,ALB,Ryanair DAC,FR8400
3,AFG,FCO,IST,2024,01,25,2024-01-25 13:05:00,Fiumicino,Havalimani,Roma,...,Turkish Airlines,TK1865,NaN,NaN,NaN,N.D.,61+,AFG,Turkish Airlines,TK1865
4,ALB,BGY,MLE,2024,02,13,FEB 13 2024,Orio al Serio,Male International,Bergamo,...,Fly Dubai,FZ1571,SEGNALATO,NaN,NaN,Permesso di soggiorno,46-60,ALB,Fly Dubai,FZ1571


In [160]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 5095 entries, 0 to 5094
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   NAZIONALITA            4979 non-null   str  
 1   AREOPORTO_ARRIVO       5095 non-null   str  
 2   AREOPORTO_PARTENZA     5095 non-null   str  
 3   ANNO_PARTENZA          5095 non-null   str  
 4   MESE_PARTENZA          5095 non-null   str  
 5   GIORNO_PARTENZA        5095 non-null   int64
 6   DATA_PARTENZA          5095 non-null   str  
 7   DESCR_AEREOPORTO_ARR   5095 non-null   str  
 8   DESCR_AEREOPORTO_PART  5095 non-null   str  
 9   CITTA_ARR              5095 non-null   str  
 10  CITTA_PARTENZA         5095 non-null   str  
 11  CODICE_PAESE_ARR       5095 non-null   str  
 12  CODICE_PAESE_PART      5095 non-null   str  
 13  PAESE_ARR              5095 non-null   str  
 14  PAESE_PART             5095 non-null   str  
 15  ZONA                   5095 non-null   str  
 16 